_keystone/oidc-federation.md

```diff
@@ -0,0 +1,503 @@
+ 1. The dropdown should show a label for the IdP. Select the IDP and click Connect.
+ 2. You will then be redirected to the IDP log in page.
+ 3. After successfully authenticating with the IDP, you should then redirected to the horizon page.
+
+## <a name="using_oauth_cli"></a>Using OAuth CLI
+using OAuth CLI for authorization code flow
+
+```
+from keystoneauth1.identity.v3 import oidc
+from keystoneauth1 import session
+from requests_oauthlib import OAuth2Session
+from keystoneauth1.identity.v3 import Token
+from novaclient import client
+import os
+
+def main():
+
+    auth_url = os.environ.get('OS_AUTH_URL')
+    identity_provider = os.environ.get('OS_IDENTITY_PROVIDER')
+    protocol = 'oidc'
+    redirect_uri = os.environ.get('OS_REDIRECT_URI')
+    scope = ['openid','profile','email']
+    client_id = os.environ.get('OS_CLIENT_ID')
+    client_secret = os.environ.get('OS_CLIENT_SECRET')
+    discovery_endpoint = os.environ.get('OS_DISCOVERY_ENDPOINT')
+    access_token_endpoint = os.environ.get('OS_TOKEN_ENDPOINT')
+    authorization_endpoint = os.environ.get('OS_AUTHORIZATION_ENDPOINT')
+    oauth = OAuth2Session(client_id, redirect_uri=redirect_uri, scope=scope)
+    authorization_url, state = oauth.authorization_url(authorization_endpoint,
+                                                       access_type='offline',
+                                                       approval_prompt='force')
+    print( 'Please go to %s and authorize access' % authorization_url)
+    authorization_code = raw_input('Enter the authorization code here: ')
+
+    s = session.Session(verify=False)
+
+    print('Got the Authorization code %s' % authorization_code)
+    print('State = %s' % state)
+
+    oidc_plugin = oidc.OidcAuthorizationCode(auth_url, identity_provider,
+                                             protocol, client_id=client_id,
+                                             client_secret=client_secret,
+                                             access_token_endpoint=access_token_endpoint,
+                                             discovery_endpoint=discovery_endpoint,
+                                             access_token_type='access_token',
+                                             redirect_uri=redirect_uri,
+                                             code=authorization_code)
+
+    unscoped_auth_token = oidc_plugin.get_unscoped_auth_ref(s)
+    scoped_token = Token(auth_url=auth_url,
+                         token=unscoped_auth_token._auth_token,
+                         project_id='<project-id>',
+                         reauthenticate=False)
+    federated_session = session.Session(auth=scoped_token)
+    nova = client.Client(version=2, session=federated_session)
+    for server in nova.servers.list():
+        print server.name
+
+if __name__ == '__main__':
+    main()
+```
+
+
+
+Using OAuth CLI for password flow
```

Comment:
You should make a note that the user credentials are never sent directly to keystone but to the oAuth provider endpoint.